### ISSUE TITLE
Too Many Login Attempts + HA Log Flooding

### ISSUE THREAD

**Issue by mgeeky** [PiotrMachowski/Home-Assistant-custom-components-Tauron-AMIplus#195]
Hi Piotr!

First of all, thank you so much for the amazing job you're doing with this (and other) integrations - AMIPlus is **an absolute blast one**! 

Many polish HA users - myself included - love reviewing Tauron eLicznik readings directly from HA dashboards. For me it became sort of a ritual to check up on them daily :)

Now, to the point, I think Tauron developers must have implemented **API calls rate restriction** lately as throughout 9 months of using your code I haven't been experiencing rate limitting, so it must be somewhat new stuff.

Can I kindly suggest implementing two features to counter described problem?

1. An additional integration configuration parameter like a "**Poll Delay**" - a number of seconds to interleave consecutive requests with
2. Refactor integration's code so that it only uses a single connection to the API, avoid authenticating more than once after integration's initialization. If re-authentication is required, wait user-configured amount of time before re-auth'ing (like for instance 10 minutes)
3. Handle "Too Many Login Attempts" extension explicitely in code and notify the user about the bug - show a log line, maybe throw HA UI dialog? Just let the user know what's going on and that adjusting config options could help.

Ad.1.
Assume Poll Delay = 3 seconds

```
- Request N
- 3 seconds
- Request N+1
- ...
```

My use case is following:

- I had to wipe out (remove and start anew) my Home Assistant database
- So I've lost my history energy readings
- When I wanted to pull previous reading with `download_statistics` service, it became stalling due to too many login attempts.

Now I'm unable to retrieve past readings as the integration **is just reading them too fast, making Tauron APIs bouncing me off**.

Exception traces:

```
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 525, in _async_step
    await getattr(self, handler)()
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 763, in _async_call_service_step
    response_data = await self._async_run_long_action(
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 726, in _async_run_long_action
    return await long_task
           ^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/core.py", line 2761, in async_call
    response_data = await coro
                    ^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/core.py", line 2804, in _execute_service
    return await target(service_call)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/tauron_amiplus/services.py", line 44, in async_handle_service
    await TauronAmiplusStatisticsUpdater.manually_update(self._hass, start_date, config_entry)
  File "/config/custom_components/tauron_amiplus/statistics.py", line 44, in manually_update
    data = await hass.async_add_executor_job(connector.get_raw_data)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/tauron_amiplus/connector.py", line 120, in get_raw_data
    data.tariff = self.login()
                  ^^^^^^^^^^^^
  File "/config/custom_components/tauron_amiplus/connector.py", line 193, in login
    session, login_response = self.login_service(CONST_URL_LOGIN, CONST_URL_SERVICE)
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/tauron_amiplus/connector.py", line 183, in login_service
    raise Exception("Too many login attempts")
Exception: Too many login attempts

[...]

Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/websocket_api/decorators.py", line 28, in _handle_async_response
    await func(hass, connection, msg)
  File "/usr/src/homeassistant/homeassistant/components/websocket_api/commands.py", line 816, in handle_execute_script
    script_result = await script_obj.async_run(
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 1795, in async_run
    return await asyncio.shield(create_eager_task(run.async_run()))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 463, in async_run
    await self._async_step(log_exceptions=False)
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 527, in _async_step
    self._handle_exception(
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 557, in _handle_exception
    raise exception
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 525, in _async_step
    await getattr(self, handler)()
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 763, in _async_call_service_step
    response_data = await self._async_run_long_action(
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 726, in _async_run_long_action
    return await long_task
           ^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/core.py", line 2761, in async_call
    response_data = await coro
                    ^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/core.py", line 2804, in _execute_service
    return await target(service_call)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/tauron_amiplus/services.py", line 44, in async_handle_service
    await TauronAmiplusStatisticsUpdater.manually_update(self._hass, start_date, config_entry)
  File "/config/custom_components/tauron_amiplus/statistics.py", line 44, in manually_update
    data = await hass.async_add_executor_job(connector.get_raw_data)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/tauron_amiplus/connector.py", line 120, in get_raw_data
    data.tariff = self.login()
                  ^^^^^^^^^^^^
  File "/config/custom_components/tauron_amiplus/connector.py", line 193, in login
    session, login_response = self.login_service(CONST_URL_LOGIN, CONST_URL_SERVICE)
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/tauron_amiplus/connector.py", line 183, in login_service
    raise Exception("Too many login attempts")
Exception: Too many login attempts
```

Another issue is that, whenever statistics collector runs, it **floods Home Assistant log** with entries like these:

```
2024-10-11 11:31:41.112 INFO (MainThread) [homeassistant.components.sensor] Setting up tauron_amiplus.sensor
2024-10-11 11:31:41.132 INFO (MainThread) [homeassistant.components.sensor] Setting up tauron_amiplus.sensor
2024-10-11 11:31:41.161 INFO (MainThread) [homeassistant.components.sensor] Setting up tauron_amiplus.sensor
2024-10-11 11:31:41.190 INFO (MainThread) [homeassistant.components.sensor] Setting up tauron_amiplus.sensor
2024-10-11 11:31:41.231 INFO (MainThread) [homeassistant.components.sensor] Setting up tauron_amiplus.sensor
2024-10-11 11:31:41.249 INFO (MainThread) [homeassistant.components.sensor] Setting up tauron_amiplus.sensor
2024-10-11 11:31:41.270 INFO (MainThread) [homeassistant.components.sensor] Setting up tauron_amiplus.sensor
2024-10-11 11:31:41.293 INFO (MainThread) [homeassistant.components.sensor] Setting up tauron_amiplus.sensor
2024-10-11 11:31:41.322 INFO (MainThread) [homeassistant.components.sensor] Setting up tauron_amiplus.sensor
2024-10-11 11:31:41.347 INFO (MainThread) [homeassistant.components.sensor] Setting up tauron_amiplus.sensor
2024-10-11 11:31:41.377 INFO (MainThread) [homeassistant.components.sensor] Setting up tauron_amiplus.sensor
```

This quickly overloads the Home Assistant server and ramps up logfile to few Megabytes. 

Is there anything you can do from the integration's API perspective to supress this? 
If not, I can always configure `logger:` section in my HA YAMLs:

```yaml
logger:
  logs:
    homeassistant.components.sensor: warning
```

Finally, can you advise would it be possible to export Energy readings in form of CSV directly from eLicznik and somehow import them to Home Assistant?

I'm thinking about manually inserting records into SQLite database, but wondering whether there's neater option :)

Keep it up Piotr, there's no HA energy dashboard for Tauron subscribers without your code! :)

**Comment by PiotrMachowski** [comment]
> Now, to the point, I think Tauron developers must have implemented API calls rate restriction lately as throughout 9 months of using your code I haven't been experiencing rate limitting, so it must be somewhat new stuff.

Actually they have added this feature quite some time ago, but problems with "Too many login attempts" usually do not happen unless HA has been restarted a few times in a short period of time. My suggestion would be to disable this integration, wait 24h and then enable it. Wait 4h and use "download statistics" functionality.

> Refactor integration's code so that it only uses a single connection to the API

This functionality is already there, except a few special cases: 
 1) This integration starts really slowly (because it loads a lot of data for some configs) - this causes HA to retry the download process. 
 2) Downloading statistics using `download_statistics` - at this moment this is a separate process which logs in by itself
 3) Downloading debug data - the same situation as with `download_statistics`

---

> Another issue is that, whenever statistics collector runs, it floods Home Assistant log with entries like these:

I haven't seen this this issue on my HA instance, I will try to catch it, but it might be related to temporarily blocked account.

---

> Finally, can you advise would it be possible to export Energy readings in form of CSV directly from eLicznik and somehow import them to Home Assistant?

I haven't tried it, but:
 * [Spook](https://spook.boo/) provides a functionality to [import statistics](https://spook.boo/recorder#import-statistics) in yaml format
 * [homeassistant-statistics](https://github.com/klausj1/homeassistant-statistics) custom integration provides a functionality to import CSV files

---

My personal solution to avoid account blocking:

I have created an automation that turns on eLicznik integration every day at 14:00 for 1h. At this moment usually data for the last day is already available. Additionally I usually restart HA in the evenings, so there this problem seems to be solved as well.

My automation:
```yaml
alias: Tauron Update
description: ""
mode: single
triggers:
  - at: "14:00:00"
    trigger: time
actions:
  - data:
      config_entry_id: ec36c11053d696303e00be209fdcb0b1
    action: homeassistant.enable_config_entry # requires Spook to be installed
  - delay:
      hours: 1
  - data:
      config_entry_id: ec36c11053d696303e00be209fdcb0b1
    action: homeassistant.disable_config_entry # requires Spook to be installed
```

**Comment by mgeeky** [comment]
Thanks Piotr for a swift reply!

Before you answered, I actually did disable integration -> waited till they unblock my account -> re-enabled and let it to run, which eventually solved my issues.

You are correct in assuming account block must have happened after series of consecutive HA restarts, that's indeed how I got here. 

Nonetheless, my issue is solved now - as I managed to rebuild statistics, supressed logs spam by reconfiguring the logger. 

Now since there's nothing from your perspective to improve code-wise, I guess my "issue" can be resolved.

Hopefully this issue and related discussion will help someone in need in future.

Thanks for your time!

Best regards,
Mariusz

**Comment by PiotrMachowski** [comment]
> Now since there's nothing from your perspective to improve code-wise, I guess my "issue" can be resolved.

Well, there are a lot of things that can be improved, but at this moment it is in "good enough" state, so it is kinda fine for me. It would take me another rewrite to change some of suboptimal behaviors of this integration, which discourages me from doing it. Maybe someday ;)

Thank you, 
Piotr
